_blogposts/compiler/overview-of-new_encoding.mdx

```diff
@@ -0,0 +1,229 @@
+
+### IC friendliness
+
+Record is always in a perfect IC position, the compiler can ensure all generated records are of the same shape
+
+## Variant (internal)
+This encoding for variants may be  subject to change in the future, but it is so simple that it makes sense for users to have a basic understanding.
+
+Take such type definition for example:
+
+```reasonml
+type t =
+  | Black(t, int, t)
+  | Red(t, int, t)
+  | Empty;
+let empty = Empty  ;
+let v0 = Black (empty, 3, empty);
+let v1 = Red (empty, 3, empty);
+```
+The generated JS code would be:
+
+```js
+var empty =/*Empty*/ 0;
+var v0  = {TAG : 0/*Black*/, _0 : /*Empty*/ 0 , _1 : 3 , _2 : /*Empty */ 0};
+var v1 = {TAG : 1/*Red*/, _0 : /*Empty*/ 0 , _1 : 3 , _2 : /*Empty */ 0};
+```
+
+As you can see, variants are divided into two categories, the variant which does not have payload is compiled into a number starting from 0, while variants which has the payload is compiled into an object which has the first slot named `TAG` and the following slots named as `_0`, `_1` ..
+
+### variant with inline records
+
+Users can give names to the payload, and the compiler respect it, however, we don't support user level renaming, i.e, using `bs.as`, at this time.
+
+```reasonml
+type t =
+  | Black ({l:t, value: int, r: t})
+  | Red({l:t, value: int, r: t})
+  | Empty;
+let empty = Empty  ;
+let v0 = Black ({l:empty, value: 3, r:empty});
+let v1 = Red ({l:empty, value:3, r: empty});
+```
+The generated JS code would be:
+```js
+var empty =/*Empty*/ 0;
+var v0  = {TAG : 0/*Black*/, l: /*Empty*/ 0 , value : 3 , r: : /*Empty */ 0};
+var v1 = {TAG : 1/*Red*/, l : /*Empty*/ 0 , value : 3 , r : /*Empty */ 0};
+``` 
+
+
+### Special case when the number of variants which has payload is only 1.
+
+Take the types below for example:
+
+```reasonml
+type list = 
+    | Nil
+    | Cons (int * list);
+```
+
+Since the number of variants which has payload is only 1, the compiler does not need add `TAG` when we destruct the data for pattern matching, so the code below:
+
+```reasonml
+let u = Cons(1,Nil)
+```
+Will generate such JS output:
+
+```js
+var u = {_0: 1, _1 : /*Nil*/ 0 }; // No TAG data.
+```
+
+### Specialized for immutable list
+
+The `list` type is a built-in type, its type definition is similar to this :
+
+```reasonml
+type t ('a) = 
+    | []
+    | (::) ('a * t ('a))
+```
+
+Without any customization,  it will generate js objects with indexes like `_0`, `_1`, since list is so pervasive,
+we provide some special treatment so that 
+```reasonml
+let u = [0,1,2,3]
+```
+Will generate js code as below:
+```js
+var u = {hd : 0, {hd : 1, {hd: 2, {hd :3 , tl : /*[]*/0 }}}}
+```
+
+This is a minor change, we changed the name of `_0` to `hd` and `_1` to `tl`.
+
+### IC friendliness
+
+For types whose number of variants which has payload is 1, it will be in a perfect IC position. 
```

Comment:
Types that have only one variant with a payload  are in a perfect position for IC.